api/src/main/java/org/apache/iceberg/VersionedSchema.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import java.io.Serializable;
+import java.util.Objects;
+
+/**
+ * A wrapper around Schema to include its schema ID
+ */
+public class VersionedSchema implements Serializable {
```

Comment:
The biggest question I have for this PR is whether we should have this wrapped class for adding schema ID, or we should just add schema ID to the `Schema` class itself, maybe as an optional `Integer`. Because Iceberg have its own jackson parser for serdes, it is not hard to ignore the `schemaId` field or set a default for it for backwards compatibility. What is the benefit of this `VersionedSchema` approach?

Reply:
I mentioned this briefly in the PR description: 

>The reason to not add ID directly into Schema is that currently schema creation is widely used as a convenient method for a lot of actions that don't involve a "real" table schema. 

The main concern I had with adding schema ID is that it might be confusing to tell from the code base when a schema should have ID and when not. A schema only needs to have ID when it's really a table schema strictly associated with table metadata, but outside of that context, the ID is not useful in any "real" operation of the schema, especially considering that it's often used for conveniently mutate fields for things like projection. Maybe we should rename `VersionedSchema` to something like `TableSchema` to better reflect this?

Reply:
I think I agree with @jackye1995 here. We could add the ID to `Schema` and just have it return `null` if there is no schema ID.

Reply:
My comment above is thinking through your question 1 in the description. I think it is okay to do this as you have it here.